File: matlab/dmp_coupling/learn_obs_avoid/utilities/computeSubFeatMatAndSubTargetCt.m
function [ varargout ] = computeSubFeatMatAndSubTargetCt( varargin )
    % Author     : Giovanni Sutanto
    % Date       : August 08, 2016
    % Description:
    %    Compute the sub-feature-matrix and the sub-target ct
    %    (as a sub-block of a matrix) from a single demonstrated 
    %    obstacle avoidance trajectory.
    %    This will later be stacked together with those 
    %    from other demonstrations, to become a huge feature matrix 
    %    (usually called X matrix) and a huge target ct 
    %    (target regression variable, usually a vector).
    
    demo_obs_avoid_traj_global              = varargin{1};
    point_obstacles_cart_position_global    = varargin{2};
    dt                                    	= varargin{3};
    cart_coord_dmp_baseline_params        	= varargin{4};
    loa_feat_methods                        = varargin{5};
    loa_feat_param                          = varargin{6};
    
    if (nargout > 2)
        if (nargout == 3)
            debugging_X_and_Ct_mode       	= 1;
        elseif (nargout == 4)
            debugging_X_and_Ct_mode         = 2;
        end
    else
        debugging_X_and_Ct_mode             = 0;
    end
    
    max_critical_point_distance_baseline_vs_oa_demo = 0.1;  % in meter
    
    Y_obs_global                            = demo_obs_avoid_traj_global{1,1};
    start_position_global_obs_avoid_demo    = Y_obs_global(1,:).';
    goal_position_global_obs_avoid_demo     = Y_obs_global(end,:).';
    
    % some error checking on the demonstration:
    % (distance between start position of baseline DMP and obstacle avoidance demonstration,
    %  as well as distance between goal position of baseline DMP and obstacle avoidance demonstration,
    %  both should be lower than max_tolerable_distance, otherwise the demonstrated obstacle avoidance
    %  trajectory is flawed):
    if ((norm(start_position_global_obs_avoid_demo-cart_coord_dmp_baseline_params.mean_start_global) > max_critical_point_distance_baseline_vs_oa_demo) || ...
        (norm(goal_position_global_obs_avoid_demo-cart_coord_dmp_baseline_params.mean_goal_global) > max_critical_point_distance_baseline_vs_oa_demo))
        disp('ERROR: Critical position distance between baseline and obstacle avoidance demonstration is beyond tolerable threshold!!!');
    end
    
    [ demo_obs_avoid_traj_local ] = convertCTrajAtOldToNewCoordSys( demo_obs_avoid_traj_global, ...
                                                                    cart_coord_dmp_baseline_params.T_global_to_local_H );
    
    [ point_obstacles_cart_position_local ]	= convertCTrajAtOldToNewCoordSys( point_obstacles_cart_position_global, ...
                                                                              cart_coord_dmp_baseline_params.T_global_to_local_H );
    
    if (debugging_X_and_Ct_mode == 2)
        [sub_X, sub_X_per_obs_point_cell]   = constructObsAvoidViconFeatMat( demo_obs_avoid_traj_local, ...
                                                                             point_obstacles_cart_position_local, ...
                                                                             dt, ...
                                                                             cart_coord_dmp_baseline_params, ...
                                                                             loa_feat_methods, ...
                                                                             loa_feat_param );
    else
        [sub_X]	= constructObsAvoidViconFeatMat( demo_obs_avoid_traj_local, ...
                                                 point_obstacles_cart_position_local, ...
                                                 dt, ...
                                                 cart_coord_dmp_baseline_params, ...
                                                 loa_feat_methods, ...
                                                 loa_feat_param );
    end
    
    Y_obs_local     = demo_obs_avoid_traj_local{1,1};
    Yd_obs_local    = demo_obs_avoid_traj_local{2,1};
    Ydd_obs_local   = demo_obs_avoid_traj_local{3,1};
    [sub_target_ct] = computeDMPCtTarget(   Y_obs_local, ...
                                            Yd_obs_local, ...
                                            Ydd_obs_local, ...
                                            cart_coord_dmp_baseline_params.w, ...
                                            cart_coord_dmp_baseline_params.n_rfs, ...
                                            cart_coord_dmp_baseline_params.mean_start_local, ...
                                            cart_coord_dmp_baseline_params.mean_goal_local, ...
                                            dt, ...
                                            cart_coord_dmp_baseline_params.c_order );
    if (debugging_X_and_Ct_mode > 0)
        sub_target_ct_3D= sub_target_ct;
    end
    
    if (strcmp(loa_feat_param.feat_constraint_mode, '_CONSTRAINED_') == 1)
        sub_target_ct   = reshape(sub_target_ct.', [(size(sub_target_ct,1)*size(sub_target_ct,2)),1]);
    end
                                   
    varargout(1)        = {sub_X};
    varargout(2)        = {sub_target_ct};
    if (debugging_X_and_Ct_mode > 0)
        varargout(3)    = {sub_target_ct_3D};
        if (debugging_X_and_Ct_mode == 2)
            varargout(4)= {sub_X_per_obs_point_cell};
        end
    end
end

File: python/dmp_coupling/learn_obs_avoid/utilities/computeSubFeatMatAndSubTargetCt.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy
import glob
sys.path.append(os.path.join(os.path.dirname(__file__), '../../../dmp_state/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../../dmp_discrete/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../../cart_dmp/cart_coord_dmp/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../../utilities/'))
from DMPState import *
from DMPTrajectory import *
from TransformSystemDiscrete import *
from CartesianCoordDMP import *
from CartesianCoordTransformer import *
from constructObsAvoidViconFeatMat import *
from DataIO import *
from utilities import *

def computeSubFeatMatAndSubTargetCt(demo_obs_avoid_traj_global, point_obstacles_cart_position_global, 
                                    dt, cart_coord_dmp_baseline_params, cart_coord_dmp):
    max_critical_point_distance_baseline_vs_oa_demo = 0.1 # in meter
    
    start_position_global_obs_avoid_demo = demo_obs_avoid_traj_global.getDMPStateAtIndex(0).getX()
    goal_position_global_obs_avoid_demo = demo_obs_avoid_traj_global.getDMPStateAtIndex(demo_obs_avoid_traj_global.getLength()-1).getX()
    
    # some error checking on the demonstration:
    # (distance between start position of baseline DMP and obstacle avoidance demonstration,
    #  as well as distance between goal position of baseline DMP and obstacle avoidance demonstration,
    #  both should be lower than max_tolerable_distance, otherwise the demonstrated obstacle avoidance
    #  trajectory is flawed):
    if ((np.linalg.norm(start_position_global_obs_avoid_demo - cart_coord_dmp_baseline_params['mean_start_global_position']) > max_critical_point_distance_baseline_vs_oa_demo) or
        (np.linalg.norm(goal_position_global_obs_avoid_demo - cart_coord_dmp_baseline_params['mean_goal_global_position']) > max_critical_point_distance_baseline_vs_oa_demo)):
        print ('ERROR: Critical position distance between baseline and obstacle avoidance demonstration is beyond tolerable threshold!!!')
        is_good_demo = False
    else:
        is_good_demo = True
    
    cart_coord_dmp.setParams(cart_coord_dmp_baseline_params['W'], cart_coord_dmp_baseline_params['A_learn'])
    
    demo_obs_avoid_traj_local = cart_coord_dmp.cart_coord_transformer.computeCTrajAtNewCoordSys(demo_obs_avoid_traj_global,
                                                                                                cart_coord_dmp_baseline_params['T_global_to_local_H'])
    
    point_obstacles_cart_position_local = cart_coord_dmp.cart_coord_transformer.computeCPosAtNewCoordSys(point_obstacles_cart_position_global.T,
                                                                                                         cart_coord_dmp_baseline_params['T_global_to_local_H'])
    point_obstacles_cart_position_local = point_obstacles_cart_position_local.T
    
    [sub_Ct_target, _, 
     sub_phase_PSI, sub_phase_X, sub_phase_V, 
     _, _, 
     _] = cart_coord_dmp.transform_sys_discrete.getTargetCouplingTermTraj(demo_obs_avoid_traj_local, 
                                                                          1.0/dt,
                                                                          cart_coord_dmp_baseline_params['mean_goal_local_position'])
    
    sub_X = constructObsAvoidViconFeatMat(demo_obs_avoid_traj_local,
                                          point_obstacles_cart_position_local,
                                          dt,
                                          cart_coord_dmp_baseline_params,
                                          cart_coord_dmp)
    
    return sub_X, sub_Ct_target, sub_phase_PSI, sub_phase_V, sub_phase_X, is_good_demo